### ISSUE TITLE
Type-narrowing doesn't work

### ISSUE THREAD

**Issue by lcswillems** [vercel/swr#3055]
# Bug report

## Description / Observed Behavior

const { data, isSuccess } = useSWR<number>('/api/user/123', fetcher);

if (isSuccess) {
  data // number | undefined
}

## Expected Behavior

It should be of type `number`.

**Comment by Ram4GB** [comment]
I don't think `useSWR` returns `isSuccess`. Are you mistyping something @lcswillems ?

**Comment by lcswillems** [comment]
@Ram4GB My bad. Fixed the example!

**Comment by Ram4GB** [comment]
I assume you request the user who is not found in your database or has been deleted by other actions. In this case, the API must throw the error with the status `400`, so you won't access your data, and it has to return `undefined`. @lcswillems

**Comment by lcswillems** [comment]
Not sure to understand. I'm just saying if "isLoading" is false, then "data" should become of type "number".

It is a basic functionality of TanStack query: https://tanstack.com/query/latest/docs/framework/react/typescript#type-narrowing

**Comment by icyJoseph** [comment]
But if fetching fails/errors, there won't be data, and it won't be loading either.

**Comment by lcswillems** [comment]
@icyJoseph Updated my example. Makes sense? Basically I'm asking for the same thing than TanStack Query.

**Comment by Ram4GB** [comment]
Yes @lcswillems, your example is not clear to me. In your case, I believe we should get the type `number` instead of `number` | `undefined`.

**Comment by lcswillems** [comment]
Yes we should get the type `number`, it is what I say in the "expected behavior". Currently, the type is `number | undefined` (Observed behavior).

**Comment by samuel871211** [comment]
@lcswillems 
I get your point.
Basically there are three ways to achieve your goal
1. 
`const { data, error, isLoading } = useSWR<number>('/api/user/123', fetcher, { suspense: true });`
2.
`const { data, error, isLoading } = useSWR<number>('/api/user/123', fetcher, { fallbackData: 0 });`
3.
```
const { data, error, isLoading } = useSWR<number>('/api/user/123', fetcher);

if (!isLoading && !error && data !== undefined) {
  data // number
}
```

**Comment by lcswillems** [comment]
Thank you for sharing!

1. Uses suspense. Don't want to go in that direction.

2. `data` would then be always a `number`. I still want it to be potentially `undefined`.

3. Yeah but the whole point of this issue is to say I should not have to add `data !== undefined`.

The 4th and best option would be to improve the typing of the library.

**Comment by samuel871211** [comment]
@lcswillems
Improve the typing of the library seems to be the best solution, but there is a bit difference between your example and TanStack query.

Your example
```
const { data, error, isLoading } = useSWR<number>('/api/user/123', fetcher);

if (!isLoading && !error) {
  data // number | undefined
}
```

TanStack query
```
const { data, isSuccess } = useQuery({
  queryKey: ['test'],
  queryFn: () => Promise.resolve(5),
})

if (isSuccess) {
  data
  //  ^? const data: number
}
```

TanStack query creates a new field `isSuccess` to determine whether type of the `data` is `Data | undefined` or `Data`

**Comment by lcswillems** [comment]
@samuel871211 In TanStack query, this also narrows down:

```
const { data, isLoading, isError } = useQuery({ ... });

if (!isLoading && !isError) {
  data
  //  ^? const data: number
}
```

And I'm wondering what prevents SWR to narrow down here:

```
const { data, error, isLoading } = useSWR<number>('/api/user/123', fetcher);

if (!isLoading && !error) {
  data // number
}
```

?

**Comment by samuel871211** [comment]
@lcswillems 

1. I'm using `"@tanstack/react-query": "^5.64.2"`

Your code example doesn't work for me.
```
const { data, isLoading, isError } = useQuery({ ... });

if (!isLoading && !isError) {
  data
  //  ^? const data: number
}
```
<img width="827" alt="Image" src="https://github.com/user-attachments/assets/568c92f4-9180-4f90-8165-ec54f31609f2" />

2. I believe SWR currently doesn't have this feature, so I have made a feature [PR](https://github.com/vercel/swr/pull/4074).

Original Codebase before [PR](https://github.com/vercel/swr/pull/4074)
```
export interface SWRResponse<Data = any, Error = any, Config = any> {
  /**
   * The returned data of the fetcher function.
   */
  data: BlockingData<Data, Config> extends true ? Data : Data | undefined
  ......
}

export type BlockingData<
  Data = any,
  Options = SWROptions<Data>
> = Options extends undefined
  ? false
  : Options extends { suspense: true }
  ? true
  : Options extends { fallbackData: Data | Promise<Data> }
  ? true
  : false
```

The type of data is `Data` only when the user provides `options.suspense` or `options.fallbackData`.

**Comment by lcswillems** [comment]
@samuel871211 This is weird because on my side it narrows down:

![Image](https://github.com/user-attachments/assets/7e27687c-7630-44e4-8ccd-10852e1657e6)

https://stackblitz.com/edit/github-j5aexizp?file=src%2Findex.tsx
